### PR TITLE
CHECKOUT-1234: Expose translation object

### PR DESCRIPTION
**Pull request by davidchin**
## What?
* Expose translated strings as a JSON object, so that it can be directly consumed by UCO (and other JS apps). i.e.:

```html
<script>
    window.language = {{{langJson 'checkout'}}};
</script>
```

```json
{
    "locale": "en-US",
    "locales": {
        "checkout.address.city_label": "en",
        "checkout.address.city_required_error": "en",
        "checkout.customer.checkout_as_guest_text": "en",
        "checkout.customer.continue_as_guest_action": "en-US"
    },
    "translations": {
        "checkout.address.city_label": "City",
        "checkout.address.city_required_error": "City is required",
        "checkout.customer.checkout_as_guest_text": "Checking out as a <strong>Guest</strong>? You'll be able to save your details to create an account with us later.",
        "checkout.customer.continue_as_guest_action": "Continue as guest"
    }
}
```
* Rename `Localizer` as `Translator` because the module only translates strings. AFAIK, `localizer` is an internal module so the change in interface should be ok.
* Improve the test coverage of the translation module. Previously, you get this warning in the coverage report `lib/localizer.js missing coverage on line(s): 21, 57, 120, 136`.
* Fix the way formatters are cached. Previously, we pass in a `translations` object as a parameter  by reference and directly assign formatters to that object. Generally speaking, we shouldn't modify parameters that are passed in by reference.
* Improve the performance of the function used to flatten the translation object. Previously, it does an extra loop which increases its complexity.
* Move some private functions, such as data transformation and filtering, into task-specific modules.
* Remove unneccessary dependency on `hoek`.

## Why?
* Currently UCO has its own translation system - `gettext`. We want to change that so it uses Stencil translation system. In order to do that, we need to expose the translated strings as a JSON object. So that it can be used to configure UCO.
* Other changes are code quality improvements.

## Testing / Proof
* Unit

@bigcommerce/checkout @mcampa @lord2800 @mjschock

### CODE REVIEW

**Comment by mcampa** [comment]
Nice refactor 👍

**Comment by davidchin** [comment]
During testing, I discover that we're not formatting strings according to their locale when we merge multiple language files. It is an existing issue affecting the master branch at the moment.

Currently, language files get cascaded - region-specific, regular and fallback (English) language files get merged in that order. In order to format these merged translations correctly, we need to have a `MessageFormat` instance for each applicable locale. But we don't have that [at the moment](https://github.com/bigcommerce/paper/blob/master/lib/localizer.js#L118).

An example of this problem:

**en.json**
```json
{ "items": "{count, plural, one{1 Item} other{# Items}}" }
```

**zh.json**
```json
{ "days": "{count, plural, other{# 天}}" }
```

So if your preferred language is `zh` (Chinese), you'll get the following result.

```html
{{lang 'days' count=1}} <!-- Prints "1 天" -->
{{lang 'items' count=1}} <!-- Prints "1 Items" -->
```

This is because Paper is using Chinese MessageFormat instance to format the fallback string, and Chinese doesn't have plurals. It only supports "other" http://www.unicode.org/cldr/charts/latest/supplemental/language_plural_rules.html

Anyway, I've made a new commit addressing the issue above. @mcampa @lord2800 @icatalina @jordanwink201, can you please take a look again?

File: lib/translator/index.js [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+/**
+ * @module paper/lib/translator
+ */
+
+const _ = require('lodash');
+const MessageFormat = require('messageformat');
+const Filter = require('./filter');
+const LocaleParser = require('./locale-parser');
+const Logger = require('../logger');
+const Transformer = require('./transformer');
```
Comment (jordanwink201) [inline lib/translator/index.js]: would be nice if these `const` were 🔤 
🍹

**Comment by jordanwink201** [comment]
good stuff :+1:

**Comment by Ubersmake** [comment]
💚

**Pull request merged**